### PR TITLE
RavenDB-18297: Implementation of Distinct provider

### PR DESCRIPTION
**Pull request by redknightlois**
In the process I have discovered a CompactTree.Iterator bug that had some ripple effects in the code, so needed to update every operation it impacted.

### CODE REVIEW

File: src/Corax/Queries/TermProviders/TermProvider.Distinct.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        private readonly Slice _term;
+        private CompactTree.Iterator _iterator;
+        
+        public DistinctTermProvider(IndexSearcher searcher, ByteStringContext context, CompactTree tree, string field, Slice term)
+        {
+            _tree = tree;
+            _searcher = searcher;
+            _field = field;
+            _iterator = tree.Iterate();
+            _iterator.Reset();
```
Comment (arekpalinski) [inline src/Corax/Queries/TermProviders/TermProvider.Distinct.cs:25]: We just created it - why doing reset?

**Comment by redknightlois** [comment]
It's the way we are doing it in general. Seek or Reset are the only
available operations after creation (which it does mostly nothing) to avoid
useless initialization code to run.

File: src/Voron/Data/CompactTrees/CompactTree.Iterator.cs [hunk 2]
```diff
@@ -57,6 +57,9 @@ public bool MoveNext(out Slice key, out long value)
                 var next = MoveNext(out Span<byte> keySpan, out value);
                 if (next)
                 {
+                    // We remove the null termination of the string to return the Slice as it is not needed by
+                    // the underlying abstraction
+                    // Slice.From(_tree._llt.Allocator, keySpan[0..^1], out key);
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.Iterator.cs]: The comment should go as well?

**Review by ayende — COMMENTED** [review]
What is the _meaning_ of this? However?
You are doing distinct on the terms, so I don't understand what the query for this will look like?

What sort of RQL will make use of this facility?

**Comment by redknightlois** [comment]
> What is the _meaning_ of this? However? You are doing distinct on the terms, so I don't understand what the query for this will look like?

Unless I didn't understood @maciejaszyk what was the actual behavior he needed for `Distinct` https://issues.hibernatingrhinos.com/issue/RavenDB-18297

**Comment by ayende** [comment]
Consider this query:

`from Users where Age > 10 select distinct Address.City`

That is the sort of query that I have in mind when we talk about `distinct`.

I'm not even sure how to express what you did as a query, but I'm fairly certain that it wouldn't actually have an impact.
Remember that the `CompactTree` has a single value per term, so iterating over that should already give you unique values.

**Comment by redknightlois** [comment]
That operation needs to be done at the select level, not at the term level. So in a sense it is a primitive like `SortingMatch`. Got it. 
EDIT: If there is no reason to scan the terms to avoid certain ones, I will scrap that part of this and close the PR.

**Comment by ayende** [comment]
The issue is that the `distinct` may be applied on values that aren't even on the index entry.
That is happening currently outside of Lucene, mind. This is implemented here:

https://github.com/ravendb/ravendb/blob/9629f5b068145877c94e210ddab587e601896c52/src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexQueryingScope.cs#L142

And should probably be handled in the exact same manner in Corax as well

**Comment by maciejaszyk** [comment]
> The issue is that the `distinct` may be applied on values that aren't even on the index entry. That is happening currently outside of Lucene, mind. This is implemented here:
> 
> https://github.com/ravendb/ravendb/blob/9629f5b068145877c94e210ddab587e601896c52/src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexQueryingScope.cs#L142
> 
> And should probably be handled in the exact same manner in Corax as well

Ok, I see a couple of scenarios:
1. (AutoMap) Term is Indexed:
-> we can do it at Corax level
2. (AutoMap) Term is not indexed:
-> we have to fetch documents and filter them

3. (Static) Field is existing only in the index
->  can do it inside Corax

4. (Static) Field is inside index but is not overwritten. Doing projection we are still getting the original term from the document. So, we've to do the work on ravendb side.
image


5. (Static) Field is inside the index and it's overwritten:
image
-> we can do it at Corax

I agree with Oren.

**Comment by ayende** [comment]
@maciejaszyk The problem with doing that in Corax is that we have analyzers that may apply, and then we have to ask what we are running distinct _on_.
Having it above Corax entirely is better. `Distinct` is also relatively rare operation, so we don't need to worry too much

**Comment by maciejaszyk** [comment]
> @maciejaszyk The problem with doing that in Corax is that we have analyzers that may apply, and then we have to ask what we are running distinct _on_. Having it above Corax entirely is better. `Distinct` is also relatively rare operation, so we don't need to worry too much

Agreed. I've talked with @redknightlois and we decided to do the work in the same manner as it's done already.

**Pull request closed without merging**